// sql-tables/dist/es6/schema/schema-guards.js
import { isObject } from '../util';
export const schemeNumeric = Object.freeze({
    Decimal: true,
    Int8: true,
    Int16: true,
    Int32: true,
    Int64: true,
    UInt8: true,
    UInt16: true,
    UInt32: true,
    UInt64: true,
    TimestampMs: true,
    TimestampS: true,
});
export const schemaNonNumeric = Object.freeze({
    Boolean: true,
    Date: true,
    Ipv4: true,
    String: true,
});
export function isSchemaNonNumeric(input) {
    return schemaNonNumeric[input] || false;
}
export function isSchemaStructFn(fn, arg) {
    if (!isObject(arg)) {
        return false;
    }
    let result = true;
    for (let i in arg) {
        if (arg.hasOwnProperty(i)) {
            if (!result) {
                break;
            }
            if (!fn(arg[i])) {
                result = false;
                break;
            }
        }
    }
    return result;
}
export function isSchemaStruct(arg) {
    return isSchemaStructFn(isSchemaStructItem, arg);
}
export function isSchemaStructStrict(arg) {
    return isSchemaStructFn(isSchemaStructProp, arg);
}
export function hasConstraints(input) {
    if (isSchemaStructProp(input)) {
        if (Array.isArray(input.constraints) && input.constraints.length) {
            return true;
        }
        return false;
    }
    return false;
}
export function hasConstraint(constraint, prop) {
    if (prop.constraints && prop.constraints.length) {
        const index = prop.constraints.indexOf(constraint);
        if (index === -1) {
            if (constraint === 'NotNull') {
                return hasConstraint('PrimaryKey', prop);
            }
            else {
                return false;
            }
        }
        else {
            return true;
        }
    }
    return false;
}
export function hasDbOnlyConstraints(prop) {
    if (hasConstraint('DbModifyOnly', prop) ||
        hasConstraint('DbInternal', prop)) {
        return true;
    }
    return false;
}
export function isSchemaNumeric(input) {
    return schemeNumeric[input] === true;
}
export function isSchemaType(input) {
    return isSchemaNumeric(input) || isSchemaNonNumeric(input);
}
export function isSchemaStructProp(arg) {
    if (arg && arg.type) {
        return true;
    }
    return false;
}
export function isSchemaStructItem(arg) {
    return isSchemaStructProp(arg) || isSchemaType(arg);
}
export function isSchemaProp(arg) {
    if (arg && isSchemaStruct(arg.struct)) {
        return true;
    }
    return false;
}
export function isSchemaRelationComposite(arg) {
    if (!Array.isArray(arg.props)) {
        return false;
    }
    if (!Array.isArray(arg.propsForeign)) {
        return false;
    }
    if (typeof arg.struct !== 'string') {
        return false;
    }
    if (arg.props.length === arg.propsForeign.length) {
        return true;
    }
    return false;
}
