// sql-tables/dist/cjs/schema/type-mappings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
exports.typeMappings = [
    {
        create: 'bigint',
        generic: 'Int64',
        infoSchema: 'bigint',
        ts: 'number',
    },
    {
        create: 'bigint',
        generic: 'UInt64',
        infoSchema: 'bigint',
        ts: 'number',
    },
    {
        create: 'boolean',
        generic: 'Boolean',
        infoSchema: 'boolean',
        ts: 'boolean',
    },
    {
        create: 'inet',
        generic: 'Ipv4',
        infoSchema: 'inet',
        ts: 'string',
    },
    {
        create: '',
        createA: 'varchar',
        generic: 'String',
        infoSchema: 'character varying',
        ts: 'string',
    },
    {
        create: 'integer',
        generic: 'Int32',
        infoSchema: 'integer',
        ts: 'number',
    },
    {
        create: 'smallint',
        generic: 'Int16',
        infoSchema: 'smallint',
        ts: 'number',
    },
    {
        create: 'integer',
        generic: 'UInt32',
        infoSchema: 'integer',
        ts: 'number',
    },
    {
        create: 'smallint',
        generic: 'UInt16',
        infoSchema: 'smallint',
        ts: 'number',
    },
    {
        create: 'timestamp default current_timestamp',
        generic: 'TimestampS',
        infoSchema: 'timestamp without time zone',
        ts: 'number',
    },
];
exports.typeMappingsByInfoSchema = util_1.arrToObj(exports.typeMappings, 'infoSchema', true);
exports.typeMappingsByGeneric = util_1.arrToObj(exports.typeMappings, 'generic');
exports.constraintMappings = [
    {
        create: '',
        createA: 'CHECK',
        generic: 'Check',
    },
    {
        create: 'NOT NULL',
        generic: 'NotNull',
    },
    {
        create: 'PRIMARY KEY',
        generic: 'PrimaryKey',
    },
    {
        create: 'UNIQUE',
        generic: 'Unique',
    },
];
exports.constraintMappingsByGeneric = util_1.arrToObj(exports.constraintMappings, 'generic');
exports.automatics = {
    'Int32': 'serial PRIMARY KEY',
    'Int64': 'bigserial PRIMARY KEY',
    'TimestampS': 'timestamp default current_timestamp',
    'UInt32': 'serial PRIMARY KEY',
    'UInt64': 'bigserial PRIMARY KEY',
};
exports.appOnlyMappings = [
    'Automatic',
    'DbModifyOnly',
    'DbInternal',
    'EncryptAppLayer',
    'EncryptDbLayer',
];
