// sql-tables/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var table_1 = require("./table");
exports.createCrud = table_1.createCrud;
exports.transactionBegin = table_1.transactionStart;
exports.transactionCommit = table_1.transactionEnd;
exports.transactionRollBack = table_1.transactionRollBack;
exports.compoundInsertOrSelectIfExists = table_1.compoundInsertOrSelectIfExists;
exports.queryStream = table_1.queryStream;
exports.insertOrSelectIfExistsObservable = table_1.insertOrSelectIfExistsObservable;
exports.insert = table_1.insert;
exports.update = table_1.update;
exports.deleteStream = table_1.deleteFrom;
exports.selectWhereStream = table_1.selectWhereStream;
exports.selectFrom = table_1.selectStream;
var schema_1 = require("./schema/schema");
exports.mutateStructIntoSchemaStructs = schema_1.mutateStructIntoSchemaStructs;
exports.strictify = schema_1.strictify;
exports.validateAndFixDatabase = schema_1.validateAndFixDatabase;
var util_1 = require("./util");
exports.deepFreeze = util_1.deepFreeze;
exports.error = util_1.error;
exports.log = util_1.log;
exports.toGtZeroIntMax = util_1.toGtZeroIntMax;
exports.toStringMax = util_1.toStringMax;
var schema_to_typescript_1 = require("./schema/schema-to-typescript");
exports.writeTsFromSchema = schema_to_typescript_1.writeTsFromSchema;
var Observable_1 = require("rxjs/Observable");
exports.Observable = Observable_1.Observable;
var Subscription_1 = require("rxjs/Subscription");
exports.Subscription = Subscription_1.Subscription;
